web/modules/custom/dpl_breadcrumb/src/Services/BreadcrumbHelper.php

```diff
@@ -0,0 +1,399 @@
+          $text,
+          ['langcode' => $langcode]
+        );
+
+        $breadcrumb_string .= "/$text";
+      }
+
+      $i++;
+    }
+
+    // If the breadcrumb does not include the current page, we want to
+    // add it manually, for the URL.
+    if (!$this->includeCurrentPage) {
+      $breadcrumb_string .= "/{$node->label()}";
+    }
+
+    return $breadcrumb_string;
+  }
+
+  /**
+   * Build the breadcrumb array.
+   *
+   * @return array<mixed>
+   *   An array of breadcrumb items, to be rendered, with text and route info.
+   */
+  public function getBreadcrumb(Node $node): array {
+    $breadcrumb = $this->getBaseBreadcrumb($node);
+    $field_name = $this->getStructureFieldName();
+
+    if ($node->hasField($field_name)) {
+      return array_merge($this->getStructureBreadcrumb($node), $breadcrumb);
+    }
+
+    if ($node->hasField('field_categories')) {
+      return array_merge($this->getCategoryBreadcrumb($node), $breadcrumb);
+    }
+
+    return $breadcrumb;
+  }
+
+  /**
+   * Build the base breadcrumb, based on possible branch references.
+   *
+   * @return array<mixed>
+   *   An array of breadcrumb items, to be rendered, with text and route info.
+   */
+  private function getBaseBreadcrumb(Node $node): array {
+    $breadcrumb = [];
+
+    $branch = NULL;
+
+    if ($node->hasField('field_branch')) {
+      $branches = $node->get('field_branch')->referencedEntities();
+      $branch = reset($branches);
+    }
+
+    if ($branch instanceof Node) {
+      $url_object = $branch->toUrl();
+
+      $breadcrumb[] = [
+        'text' => $branch->label(),
+        'route_name' => $url_object->getRouteName(),
+        'route_parameters' => $url_object->getRouteParameters(),
+      ];
+    }
+
+    if ($node->bundle() === 'article') {
+      $breadcrumb[] = [
+        'text' => $this->translation->translate('Articles'),
+        'route_name' => 'view.articles.page_1',
+        'route_parameters' => [],
+      ];
+    }
+
+    return $breadcrumb;
+  }
+
+  /**
+   * Build a breadcrumb array, based on field_categories.
+   *
+   * @return array<mixed>
+   *   An array of breadcrumb items, to be rendered, with text and route info.
+   */
+  public function getCategoryBreadcrumb(Node $node): array {
+    $breadcrumb = $this->includeCurrentPage ? [
+      [
+        'text' => $node->label(),
+        'route_name' => 'entity.node.canonical',
+        'route_parameters' => ['node' => $node->id()],
+      ],
+    ] : [];
+
+    $categories = $node->get('field_categories')->referencedEntities();
+
+    $category = reset($categories);
+
+    if ($category instanceof TermInterface) {
+      $category_id = intval($category->id());
+      // Get all parent categories, including current.
+      $categories = $this->entityTypeManager->getStorage("taxonomy_term")
+        ->loadAllParents($category_id);
+
+      foreach ($categories as $category) {
+        $breadcrumb[] = [
+          'text' => $category->getName(),
+        ];
+      }
```

Comment:
because we have no "term page" frontend ready yet.
i'll add a code comment here, but it is on purpose that these breadcrumbs arent clickable right now.

Reply:
We do not but we do have a content link on the terms which we redirect to.

I guess we are not guaranteed to have a page 😕 .